### PR TITLE
Fds 1843 use graph

### PR DESCRIPTION
**Pull request by afwillia**
Address [FDS-1843](https://sagebionetworks.jira.com/browse/FDS-1843). The purpose of this PR is to update all functions that accept a jsonld schema to also accept a pickle version of the graph. When a graph is supplied instead of a schema, the functions will skip the time-consuming step of reading the schema into a graph.

[FDS-1843]: https://sagebionetworks.jira.com/browse/FDS-1843?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by linglp** [comment]
@afwillia Thanks for the PR. Please add PR description, including linking Jira issue with this PR and making sure that tests are passing. Also, I am not seeing any file changes related to tests, so please also change tests to make sure that we are testing data_model_graph_pickle as an additional parameter.

File: schematic/manifest/generator.py [hunk 1]
```diff
@@ -1633,11 +1634,14 @@ def create_manifests(
         title: Optional[str] = None,
         strict: Optional[bool] = True,
         use_annotations: Optional[bool] = False,
```
Comment (andrewelamb) [inline schematic/manifest/generator.py:1636]: Should path_to_data_model now be optional as well?

File: schematic/manifest/generator.py [hunk 2]
```diff
@@ -1633,11 +1634,14 @@ def create_manifests(
...
+        graph_data_model: Optional[nx.MultiDiGraph] = None,
+        data_model_graph_pickle: str = None,
     ) -> Union[List[str], List[pd.DataFrame]]:
         """Create multiple manifests
 
         Args:
             path_to_data_model (str): str path to data model
+            data_model_graph_pickle (str): A data model graph as csv or pickle
```
Comment (andrewelamb) [inline schematic/manifest/generator.py:1644]: Looks like you are missing a description for `graph_data_model`

File: schematic/models/metadata.py [hunk 3]
```diff
@@ -39,6 +40,7 @@ def __init__(
         inputMModelLocation: str,
         inputMModelLocationType: str,
         data_model_labels: str,
+        data_model_graph_pickle: Optional[str] = None,
```
Comment (andrewelamb) [inline schematic/models/metadata.py:43]: Why does this method only take the pickle file, where the one above takes the pickle file and a graph object?

**Comment by andrewelamb** [comment]
@afwillia It looks liky you are missing tests for `generator.create_manifests`: using both the graph object and pickle file

File: tests/test_viz.py [hunk 4]
```diff
@@ -17,10 +17,12 @@
 def attributes_explorer(helpers):
     # Get JSONLD file path
     path_to_jsonld = helpers.get_data_path("example.model.jsonld")
+    path_to_graph = helpers.get_data_path("example.model.pickle")
```
Comment (andrewelamb) [inline tests/test_viz.py:21]: I don't quite understand whats going on here. Shouldn't you be creating a 2nd attributes explorer that uses the pickle file and then testing both the original and new object?

Comment (andrewelamb) [reply tests/test_viz.py:21]: I'm not seeing any new tests for the tangles tree object.

File: tests/test_viz.py [hunk 5]
```diff
@@ -17,10 +17,12 @@
...
```
Comment (GiaJordan) [inline tests/test_viz.py:20]: If `AttributedExplorer` can accept inputs of the jsonld, the pickle, and both, then the tests should cover intitializing the objecct with those combinations. Explore parameterization for swapping the objects in place of each other for the tests/fixtures

File: schematic/manifest/generator.py [hunk 6]
```diff
@@ -1642,11 +1643,15 @@ def create_manifests(
         title: Optional[str] = None,
         strict: Optional[bool] = True,
         use_annotations: Optional[bool] = False,
+        graph_data_model: Optional[nx.MultiDiGraph] = None,
+        data_model_graph_pickle: str = None,
```
Comment (BryanFauble) [inline schematic/manifest/generator.py:1647]:
```suggestion
        data_model_graph_pickle: Optional[str] = None,
```

File: schematic/manifest/generator.py [hunk 7]
```diff
@@ -1677,16 +1682,25 @@ def create_manifests(
                     "Please check your submission and try again."
                 )
 
-        data_model_parser = DataModelParser(path_to_data_model=path_to_data_model)
+        if not graph_data_model:
+            if data_model_graph_pickle:
+                with open(data_model_graph_pickle, "rb") as f:
```
Comment (BryanFauble) [inline schematic/manifest/generator.py:1687]: Is this file always guaranteed to fit into memory? Really large files are likely to cause OOM issues here. Likely you are not dealing with large amounts of data, but thought I'd throw the question out there.

File: schematic/visualization/attributes_explorer.py [hunk 8]
```diff
@@ -26,10 +28,15 @@ def __init__(
...
         parsed_data_model: Optional[dict] = None,
+        graph_data_model: Optional[nx.MultiDiGraph] = None,
+        data_model_graph_pickle: Optional[str] = None,
     ) -> None:
         self.path_to_jsonld = path_to_jsonld
 
         self.jsonld = load_json(self.path_to_jsonld)
+        if data_model_graph_pickle and not graph_data_model:
+            with open(data_model_graph_pickle, "rb") as file:
+                graph_data_model = pickle.load(file)
```
Comment (BryanFauble) [inline schematic/visualization/attributes_explorer.py:39]: If I am reading this code correctly, doesn't this not accomplish anything?

This is being set on L53
`self.graph_data_model = data_model_grapher.graph`


On this L39 you are updating the attribute passed in if None:
`graph_data_model = pickle.load(file)`

But then `graph_data_model` is not actually being used anywhere to set anything on the `AttributesExplorer` class instance that was instantiated.

File: schematic/manifest/generator.py (hunk 1 above)
Comment (afwillia) [reply schematic/manifest/generator.py:1645]: My understanding is that we still need the jsonld data model for other operations

File: schematic/manifest/generator.py (hunk 7 above)
Comment (afwillia) [reply schematic/manifest/generator.py:1687]: I think the largest of these files will be <100MB so I don't think we need to worry about it for this release.

File: schematic/visualization/tangled_tree.py [hunk 9]
```diff
@@ -59,19 +61,27 @@ def __init__(
         # Parse schema name
         self.schema_name = path.basename(self.path_to_json_ld).split(".model.jsonld")[0]
 
+        parsed_data_model = None
+
         # Instantiate Data Model Parser
-        data_model_parser = DataModelParser(
-            path_to_data_model=self.path_to_json_ld,
-        )
+        if not data_model_graph_pickle:
```
Comment (thomasyu888) [inline schematic/visualization/tangled_tree.py:67]: Nit: Avoid doing this unless `data_model_graph_pickle` is a boolean. 

```
if data_model_graph_pickle is not None
```

File: schematic/visualization/attributes_explorer.py (hunk 8 above)
Comment (afwillia) [reply schematic/visualization/attributes_explorer.py:39]: Nice catch, thanks. I updated this to set self.graph_data_model on L37 and inside `if not data_model_grapher` on L47

File: schematic/manifest/generator.py [hunk 10]
```diff
@@ -1668,16 +1672,25 @@ def create_manifests(
                     "Please check your submission and try again."
                 )
 
-        data_model_parser = DataModelParser(path_to_data_model=path_to_data_model)
+        if not graph_data_model:
```
[the remark this answers is not in the document]
Comment (afwillia) [reply schematic/manifest/generator.py:1685]: Done!

File: schematic/models/metadata.py (hunk 3 above)
Comment (afwillia) [reply schematic/models/metadata.py:46]: My impression is that this function would read from a file, not a graph object.

**Comment by afwillia** [comment]
I appreciate everyone's feedback and have incorporated most of it into this PR. @andrewelamb @linglp do you think this looks ready to merge? I have not addressed @BryanFauble 's comment about creating more readable tests, but I think the coverage is there.

File: schematic/manifest/generator.py [hunk 11]
```diff
@@ -1642,11 +1643,15 @@ def create_manifests(
...
         use_annotations: Optional[bool] = False,
+        graph_data_model: Optional[nx.MultiDiGraph] = None,
+        data_model_graph_pickle: Optional[str] = None,
     ) -> Union[List[str], List[pd.DataFrame]]:
         """Create multiple manifests
 
         Args:
             path_to_data_model (str): str path to data model
+            data_model_graph_pickle (str): A data model graph as csv or pickle
+            graph_data_model (str): An networkx MultiDiGraph object
```
Comment (andrewelamb) [inline schematic/manifest/generator.py:1654]: This documentation doesn't match the parameter

File: schematic/manifest/generator.py [hunk 12]
```diff
@@ -1677,16 +1682,25 @@ def create_manifests(
                     "Please check your submission and try again."
                 )
 
-        data_model_parser = DataModelParser(path_to_data_model=path_to_data_model)
+        if not graph_data_model:
```
Comment (andrewelamb) [inline schematic/manifest/generator.py:1685]: You might want this to be `if graph_data_model is not None` I'm not sure how `nx.MultiDiGraph` handles thruthyness.

**Comment by thomasyu888** [comment]
Closing due to other PR

**Pull request closed without merging**